README.md

```diff
@@ -198,289 +189,494 @@ e.increment();  // logs 1
+## Possible extensions
+
+Decorators on further constructs are investigated in [EXTENSIONS.md](./EXTENSIONS.md).
+
+# Syntax
 
-## User-defined Decorators
+This decorators proposal uses the syntax of the previous Stage 2 decorators proposal. This means that:
+- Decorator expressions are restricted to a chain of variables, property access with `.` but not `[]`, and calls `()`. To use an arbitrary expression as a decorator, `@(expression)` is an escape hatch.
+- Class expressions may be decorated, not just class declarations.
+- Class decorators come after `export` and `default`.
 
-JavaScript programmers can make their own decorators by composing built-in decorators.
+There is no special syntax for defining decorators; any function can be applied as a decorator.
 
-### `decorator @xyz` declarations
+# Detailed semantics 
 
-Decorators may be defined as a simple composition of other decorators. You can use all the fancy JavaScript features you want inside the arguments, but at the top level, this is just a string of decorators and arguments for these decorators. There's no way to conditionally use one decorator in one situation and another in another situation, for example.
+The three steps of decorator evaluation:
+1. Decorator expressions (the thing after the `@`) are *evaluated* interspersed with computed property names.
+1. Decorators are *called* (as functions) during class definition, after the methods have been evaluated but before the constructor and prototype have been put together.
+1. Decorators are *applied* (mutating the constructor and prototype) all at once, after all of them have been called.
 
-Example:
+The semantics here generally follow the consensus at the May 2016 TC39 meeting in Munich.
+
+## 1. Evaluating decorators
+
+
+
+Decorators are evaluated as expressions, being ordered along with computed property names. This goes left to right, top to bottom. The result of decorators is stored in the equivalent of local variables to be later called after the class definition initially finishes executing.
+
+## 2. Calling decorators
+
+### The element being wrapped: the first parameter
+
+The first parameter, of what the decorator is wrapping, depends on what is being decorated:
+- In a method, init-method, getter or setter decorator: the relevant function object
+- In a class decorator: the class
+- In a field: An object with two properties
+    - `get`: A function which takes no arguments, expected to be called with a receiver which is the appropriate object, returning the underlying value.
+    - `set`: A function which takes a single argument (the new value), expected to be called with a receiver which is the object being set, expected to return `undefined`.
+
+### The context object: the second parameter
+
+The context object--the object passed as the second argument to the decorator--contains the following properties:
+- `kind`: One of
+    - `"class"`
+    - `"method"`
+    - `"init-method"`
```

Comment:
BTW, why not to use `isInit` property by analogy with `isStatic` instead of `init-method` approach?

Reply:
I feel like the kind should tell you which type of data to expect in the first argument.